`<ranges>`: using `views::reverse` on `ranges::reverse_view` lvalue is broken

### Describe the bug

Using `views::reverse` on (possibly const) lvalue of type `ranges::reverse_view<X>` does not result in `X`, but in `ranges::reverse_view<X>`.

### Command-line test case

```c++
#include <ranges>

using namespace std;

int main() {
    using X = ranges::iota_view<int, int>;
    ranges::reverse_view<X> view;
    static_assert( same_as<decltype(view | views::reverse), X>);
    static_assert(!same_as<decltype(view | views::reverse), ranges::reverse_view<X>>);
}
```

```text
C:\Temp>cl /std:c++latest /EHsc broken-reverse.cpp
Microsoft (R) C/C++ Optimizing Compiler Version 19.29.30136 for x86
Copyright (C) Microsoft Corporation.  All rights reserved.

/std:c++latest is provided as a preview of language features from the latest C++
working draft, and we're eager to hear about bugs and suggestions for improvements.
However, note that these features are provided as-is without support, and subject
to changes or removal as the working draft evolves. See
https://go.microsoft.com/fwlink/?linkid=2045807 for details.

broken-reverse.cpp
broken-reverse.cpp(8): error C2607: static assertion failed
broken-reverse.cpp(9): error C2607: static assertion failed
```

[Test case in compiler explorer.](https://godbolt.org/z/rqYhMaxaE)

### Expected behavior

This code should compile.

### STL version

16.11.5 and 17.0.0.

### Additional context

This code can be compiled with [GCC](https://godbolt.org/z/cv63ddxob) and [clang](https://godbolt.org/z/5TvE6baPM).

[Extra test (GCC in compiler explorer).](https://godbolt.org/z/8q67aMnW8)

--- a/stl/inc/ranges
+++ b/stl/inc/ranges
@@ -4167,24 +4167,24 @@ namespace ranges {
 
             template <class _Rng>
             _NODISCARD static _CONSTEVAL _Choice_t<_St> _Choose() noexcept {
-                if constexpr (bidirectional_range<_Rng>) {
-                    if constexpr (_Is_specialization_v<_Rng, reverse_view>) {
-                        if constexpr (_Can_extract_base<_Rng>) {
-                            return {_St::_Base, noexcept(_STD declval<_Rng>().base())};
-                        }
-                    } else if constexpr (_Reversed_subrange<remove_cvref_t<_Rng>> == 0) {
-                        using _It = decltype(_STD declval<_Rng&>().begin().base());
-                        return {_St::_Subrange_unsized,
-                            noexcept(subrange<_It, _It, subrange_kind::unsized>{
-                                _STD declval<_Rng&>().end().base(), _STD declval<_Rng&>().begin().base()})};
-                    } else if constexpr (_Reversed_subrange<remove_cvref_t<_Rng>> == 1) {
-                        using _It = decltype(_STD declval<_Rng&>().begin().base());
-                        return {_St::_Subrange_sized,
-                            noexcept(subrange<_It, _It, subrange_kind::sized>{_STD declval<_Rng&>().end().base(),
-                                _STD declval<_Rng&>().begin().base(), _STD declval<_Rng&>().size()})};
-                    } else if constexpr (_Can_reverse<_Rng>) {
-                        return {_St::_Reverse, noexcept(reverse_view{_STD declval<_Rng>()})};
+                using _Ty = remove_cvref_t<_Rng>;
+
+                if constexpr (_Is_specialization_v<_Ty, reverse_view>) {
+                    if constexpr (_Can_extract_base<_Rng>) {
+                        return {_St::_Base, noexcept(_STD declval<_Rng>().base())};
                     }
+                } else if constexpr (_Reversed_subrange<_Ty> == 0) {
+                    using _It = decltype(_STD declval<_Rng&>().begin().base());
+                    return {_St::_Subrange_unsized,
+                        noexcept(subrange<_It, _It, subrange_kind::unsized>{
+                            _STD declval<_Rng&>().end().base(), _STD declval<_Rng&>().begin().base()})};
+                } else if constexpr (_Reversed_subrange<_Ty> == 1) {
+                    using _It = decltype(_STD declval<_Rng&>().begin().base());
+                    return {_St::_Subrange_sized,
+                        noexcept(subrange<_It, _It, subrange_kind::sized>{_STD declval<_Rng&>().end().base(),
+                            _STD declval<_Rng&>().begin().base(), _STD declval<_Rng&>().size()})};
+                } else if constexpr (_Can_reverse<_Rng>) {
+                    return {_St::_Reverse, noexcept(reverse_view{_STD declval<_Rng>()})};
                 }
 
                 return {_St::_None};

--- a/tests/std/tests/P0896R4_views_reverse/test.cpp
+++ b/tests/std/tests/P0896R4_views_reverse/test.cpp
@@ -317,6 +317,13 @@ using move_only_view = test::range<Category, const int, test::Sized{is_random}, 
     IsCommon, test::CanCompare{derived_from<Category, forward_iterator_tag>},
     test::ProxyRef{!derived_from<Category, contiguous_iterator_tag>}, test::CanView::yes, test::Copyability::move_only>;
 
+void test_gh2312() {
+    using X = ranges::iota_view<int, int>;
+    ranges::reverse_view<X> view;
+    static_assert( same_as<decltype(view | views::reverse), X>);
+    static_assert(!same_as<decltype(view | views::reverse), ranges::reverse_view<X>>);
+}
+
 int main() {
     static constexpr int some_ints[]     = {0, 1, 2};
     static constexpr int reversed_ints[] = {2, 1, 0};

--- a/tests/std/tests/P0896R4_views_reverse/test.cpp
+++ b/tests/std/tests/P0896R4_views_reverse/test.cpp
@@ -317,7 +317,7 @@ using move_only_view = test::range<Category, const int, test::Sized{is_random}, 
     IsCommon, test::CanCompare{derived_from<Category, forward_iterator_tag>},
     test::ProxyRef{!derived_from<Category, contiguous_iterator_tag>}, test::CanView::yes, test::Copyability::move_only>;
 
-void test_gh2312() {
+void test_gh2312() { // COMPILE-ONLY
     using X = ranges::iota_view<int, int>;
     ranges::reverse_view<X> view;
     static_assert(same_as<decltype(view | views::reverse), X>);

--- a/tests/std/tests/P0896R4_views_reverse/test.cpp
+++ b/tests/std/tests/P0896R4_views_reverse/test.cpp
@@ -320,8 +320,7 @@ using move_only_view = test::range<Category, const int, test::Sized{is_random}, 
 void test_gh2312() {
     using X = ranges::iota_view<int, int>;
     ranges::reverse_view<X> view;
-    static_assert( same_as<decltype(view | views::reverse), X>);
-    static_assert(!same_as<decltype(view | views::reverse), ranges::reverse_view<X>>);
+    static_assert(same_as<decltype(view | views::reverse), X>);
 }
 
 int main() {

